pkg/gatewayconfigurationserver/server.go

```diff
@@ -0,0 +1,159 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gatewayconfigurationserver
+
+import (
+	"context"
+	"encoding"
+	"encoding/json"
+	"net/http"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/gorilla/mux"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"go.thethings.network/lorawan-stack/v3/pkg/component"
+	gcsv2 "go.thethings.network/lorawan-stack/v3/pkg/gatewayconfigurationserver/v2"
+	"go.thethings.network/lorawan-stack/v3/pkg/pfconfig/cpf"
+	"go.thethings.network/lorawan-stack/v3/pkg/pfconfig/semtechudp"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/web"
+	"go.thethings.network/lorawan-stack/v3/pkg/webhandlers"
+	"google.golang.org/grpc"
+)
+
+// Server implements the Gateway Configuration Server component.
+type Server struct {
+	*component.Component
+	config *Config
+}
+
+func (s *Server) withGateway(next func(http.ResponseWriter, *http.Request, *ttnpb.Gateway)) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		gtwID := r.Context().Value(gatewayIDKey).(ttnpb.GatewayIdentifiers)
+		cc, err := s.GetPeerConn(r.Context(), ttnpb.ClusterRole_ENTITY_REGISTRY, nil)
```

Comment:
If you pass the gatewayID to GetPeerConn, we might hit the same IS peer on multiple calls and benefit from local caching